meilisearch-core/src/query_builder.rs

```diff
@@ -88,97 +138,25 @@ impl<'c, 'f, 'd, 'i> QueryBuilder<'c, 'f, 'd, 'i> {
         range: Range<usize>,
     ) -> MResult<SortResult> {
-        let facets_docids = match self.facet_filter {
-            Some(facets) => {
-                let mut ands = Vec::with_capacity(facets.len());
-                let mut ors = Vec::new();
-                for f in facets.deref() {
-                    match f {
-                        Either::Left(keys) => {
-                            ors.reserve(keys.len());
-                            for key in keys {
-                                let docids = self.index.facets.facet_document_ids(reader, &key)?.unwrap_or_default();
-                                ors.push(docids);
-                            }
-                            let sets: Vec<_> = ors.iter().map(Cow::deref).collect();
-                            let or_result = sdset::multi::OpBuilder::from_vec(sets).union().into_set_buf();
-                            ands.push(Cow::Owned(or_result));
-                            ors.clear();
-                        }
-                        Either::Right(key) =>{
-                            match self.index.facets.facet_document_ids(reader, &key)? {
-                                Some(docids) => ands.push(docids),
-                                // no candidates for search, early return.
-                                None => return Ok(SortResult::default()),
-                            }
-                        }
-                    };
-                }
-                let ands: Vec<_> = ands.iter().map(Cow::deref).collect();
-                Some(sdset::multi::OpBuilder::from_vec(ands).intersection().into_set_buf())
-            }
-            None => None
-        };
 
-        // for each field to retrieve the count for, create an HashMap associating the attribute
-        // value to a set of matching documents. The HashMaps are them collected in another
-        // HashMap, associating each HashMap to it's field.
-        let facet_count_docids = match self.facets {
-            Some(field_ids) => {
-                let mut facet_count_map = HashMap::new();
-                for (field_id, field_name) in field_ids {
-                    let mut key_map = HashMap::new();
-                    for pair in self.index.facets.field_document_ids(reader, field_id)? {
-                        let (facet_key, document_ids) = pair?;
-                        let value = facet_key.value();
-                        key_map.insert(value.to_string(), document_ids);
-                    }
-                    facet_count_map.insert(field_name, key_map);
-                }
-                Some(facet_count_map)
+        // When there is no candidate document in the facet filter, we don't perform the search,
+        // as the result will alway be empty.
+        if let Some(ref facet_filter) = self.facet_filter {
+            if facet_filter.is_empty() {
+                return Ok(SortResult::default())
             }
-            None => None,
-        };
+        }
```

Comment:
I'm not *super* happy with this, but this is a trade-off that's required for performance.